internal/osbuild/gcp_guest_agent_conf_stage.go

```diff
@@ -0,0 +1,104 @@
+package osbuild
+
+import (
+	"fmt"
+	"regexp"
+)
+
+const configScopeRegex = "^(distro|instance)$"
+
+type Accounts struct {
+	DdeprovisionRemove *bool    `json:"deprovision_remove,omitempty"`
+	Groups             []string `json:"groups,omitempty"`
+	UseraddCmd         string   `json:"useradd_cmd,omitempty"`
+	UserdelCmd         string   `json:"userdel_cmd,omitempty"`
+	UsermodCmd         string   `json:"usermod_cmd,omitempty"`
+	GpasswdAddCmd      string   `json:"gpasswd_add_cmd,omitempty"`
+	GpasswdRemoveCmd   string   `json:"gpasswd_remove_cmd,omitempty"`
+	GroupaddCmd        string   `json:"groupadd_cmd,omitempty"`
+}
+
+type Daemons struct {
+	AccountsDaemon  *bool `json:"accounts_daemon,omitempty"`
+	ClockSkewDaemon *bool `json:"clock_skew_daemon,omitempty"`
+	NetworkDaemon   *bool `json:"network_daemon,omitempty"`
+}
+
+type InstanceSetup struct {
+	HostKeyTypes     []string `json:"host_key_types,omitempty"`
+	OptimizeLocalSsd *bool    `json:"optimize_local_ssd,omitempty"`
+	NetworkEnabled   *bool    `json:"network_enabled,omitempty"`
+	SetBotoConfig    *bool    `json:"set_boto_config,omitempty"`
+	SetHostKeys      *bool    `json:"set_host_keys,omitempty"`
+	SetMultiqueue    *bool    `json:"set_multiqueue,omitempty"`
+}
+
+type IpForwarding struct {
+	EthernetProtoId   string `json:"ethernet_proto_id,omitempty"`
+	IpAliases         *bool  `json:"ip_aliases,omitempty"`
+	TargetInstanceIps *bool  `json:"target_instance_ips,omitempty"`
+}
+
+type MetadataScripts struct {
+	DefaultShell string `json:"default_shell,omitempty"`
+	RunDir       string `json:"run_dir,omitempty"`
+	Startup      *bool  `json:"startup,omitempty"`
+	Shutdown     *bool  `json:"shutdown,omitempty"`
+}
+
+type NetworkInterfaces struct {
+	Setup        *bool  `json:"setup,omitempty"`
+	IpForwarding *bool  `json:"ip_forwarding,omitempty"`
+	DhcpCommand  string `json:"dhcp_command,omitempty"`
+}
+
+type Config struct {
```

Comment:
For all of the exported structures in this file, I suggest to prefix their names with some part of the stage name (e.g. `GcpGuestAgentConfig`). Otherwise, they'll get exported as `osbuild.<struct_name>` and it is really not obvious to which osbuild stage they are related. With the `Config` structure it gets probably as bad as it can, because it gets used as `osbuild.Config{}`, which looks as if it was some configuration for osbuild itself.

Although the structure names will get longer, it is IMO well worth the increased code readability.